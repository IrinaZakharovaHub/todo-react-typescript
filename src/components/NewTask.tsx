import * as React from 'react';
import {TaskName} from "../types/taskType";

interface Props {
    changeTask: (event: React.ChangeEvent<HTMLInputElement>) => void;
    addTask: (event: React.FormEvent<HTMLFormElement>, task: TaskName) => void;
    taskName: TaskName;
}

const NewTask: React.FC<Props> = ({taskName, addTask, changeTask}) => {
    return (
        <>
            <form onSubmit={(event)=>addTask(event, taskName)}>
                <input type='text' onChange={(event)=>changeTask(event)} value={taskName}/>
                <button type="submit">Add a task</button>
            </form>
        </>
    )
};

export default NewTask;
