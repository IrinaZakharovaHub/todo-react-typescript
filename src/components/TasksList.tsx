import * as React from 'react';

import {Task, Tasks} from "../types/taskType";

interface Props {
    tasks: Tasks;
    removeTask: (taskForRemoving: Task) => void;
    toggleReadiness: (taskForChange: Task) => void;
}

const TasksList: React.FC<Props> = ({tasks, removeTask, toggleReadiness}) => {
    return (
        <>
            <ul>
                {tasks.map((task,i)=>(
                    <li key={i} className={task.isDone ? 'ready' : null}>
                        <label>
                            <input type="checkbox" onChange={()=>toggleReadiness(task)} checked={task.isDone}/>
                        </label>
                        {task.name}
                        <button className='remove-button' onClick={()=>removeTask(task)}>
                            X
                        </button>
                    </li>
                ))}
            </ul>
        </>
    )
};

export default TasksList;
