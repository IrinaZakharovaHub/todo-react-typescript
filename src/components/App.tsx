import * as React from 'react';
import {useState} from "react";

import NewTask from "./NewTask";
import TasksList from "./TasksList";

import {TaskName, Task, Tasks} from "../types/taskType";


const App = () => {
    const [newTaskName, setTaskName] = useState<TaskName>('');
    const [tasks, setTasks] = useState<Tasks>([]);

    const addTask = (event: React.FormEvent<HTMLFormElement>, task: TaskName) => {
        event.preventDefault();
        setTasks([...tasks, {
            name: task,
            isDone: false
        }]);
        setTaskName('');
    }

    const changeTask = (event: React.ChangeEvent<HTMLInputElement>) => {
        setTaskName(event.target.value);
    }

    const removeTask = (taskForRemoving: Task) => {
        setTasks(previousTasks => (
            [...previousTasks.filter(task => task.name !== taskForRemoving.name)]
        ))
    }
    const toggleReadiness = (taskForChange: Task) => {
        setTasks(previousTasks => (
            [...previousTasks.map((task) => (task !== taskForChange ? task : {...task, isDone: !task.isDone}))]
        ))
    }

    return (
        <>
            <div>
                <NewTask
                    taskName={newTaskName}
                    addTask={addTask}
                    changeTask={changeTask}
                />
                <TasksList
                    tasks={tasks}
                    removeTask={removeTask}
                    toggleReadiness={toggleReadiness}
                />
            </div>
        </>
    )
}

export default App;
