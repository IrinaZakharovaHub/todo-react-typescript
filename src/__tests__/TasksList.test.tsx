import * as React from 'react';
import {ContextApp} from "../components/App";

import {shallow} from "enzyme";
import {render} from "@testing-library/react";
import {renderHook } from '@testing-library/react-hooks';

import {useReducer} from "react";
import {todoReducer} from "../reducers/todoReducer";

import TasksList from "../components/TasksList";

describe('<TasksList />',() => {

    const testState = {
        newTask: '',
        tasks: [{name: 'test', isDone: false}, {name: 'test2', isDone: false}]
    }

    const { result } = renderHook(()=> useReducer(todoReducer, testState));
    const [state, changeState] = result.current;

    const wrapper =
        <ContextApp.Provider value={{state, changeState}}>
            <TasksList/>
        </ContextApp.Provider>;

    it('renders the component without changes', () => {
        const component = shallow(wrapper);
        expect(component).toMatchSnapshot();
    });

    it('should render right input value', () => {
        const {container} = render(wrapper);
        expect(container.querySelectorAll('li')).toHaveLength(testState.tasks.length);
    });

})
