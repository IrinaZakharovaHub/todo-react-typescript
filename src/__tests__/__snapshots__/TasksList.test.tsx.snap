// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TasksList /> renders the component without changes 1`] = `
<ContextProvider
  value={
    Object {
      "state": Object {
        "newTask": "",
        "tasks": Array [
          Object {
            "isDone": false,
            "name": "test",
          },
          Object {
            "isDone": false,
            "name": "test2",
          },
        ],
      },
    }
  }
>
  <TasksList />
</ContextProvider>
`;
